DCOS-595: Stop writing zookeeper log to a file

mhrabovcin opened the pull request:
## High Level Description

Currently Zookeeper writes its logs to the `/var/lib/dcos/exhibitor/zookeeper/zookeeper.out` file and also to the `journald` service. Since the file isn't managed by any logrotation service it grows unbounded and can fill up disk. This patch links `/var/lib/dcos/exhibitor/zookeeper/zookeeper.out` to a `/dev/null` to prevent filling up disk. Users should use `journald` service to examine `zookeeper` and `exhibitor` logs. Part of the patch is also setting explicitly where the log file exists so it won't happen that location will change in future releases without controlling it.

## Related Issues

  - [DCOS-595](https://dcosjira.atlassian.net/browse/DCOS-595) DC/OS Zookeeper does not log manage.

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]
___
**PLEASE FILL IN THE TEMPLATE ABOVE** / **DO NOT REMOVE ANY SECTIONS ABOVE THIS LINE**

## Instructions and Review Process

**What is the review process and when will my changes land?**

All PRs require 2 approvals using github's [pull request reviews](https://help.github.com/articles/about-pull-request-reviews/).

Reviewers should be:
* Developers who understand the code being modified.
* Developers responsible for code that interacts with or depends on the code being modified.

It is best to proactively ask for 2 reviews by @mentioning the candidate reviewers in the PR comments area. The responsibility is on the developer submitting the PR to follow-up with reviewers and make sure a PR is reviewed in a timely manner. Once a PR has **2 ship-it's**, **no red reviews**, and **all tests are green** it will be included in the [next train](https://github.com/dcos/dcos/blob/master/contributing.md).

lingmann commented:
@mhrabovcin can you update the checklist items underneath "Checklist for all PR's"? If there are no tests for this PR, please just include reasoning as to why it's not justified/necessary. Thanks.

lingmann commented:
So it looks like this PR has failed twice on our Azure integration tests... and I suspect the cluster is failing to come up properly on Azure with these changes. @mhrabovcin can you investigate further? If you search the build logs for `For troubleshooting, master0 can be reached using` you'll find some details on how to SSH to a master node. Alternatively, if you rebase and get @mellenburg 's latest Azure stuff you'll need to search for a slightly different line... but he can point you in the right direction. Also feel free to reach out on #foundations channel if you need additional help to connect to these Azure instances.

mhrabovcin commented:
@lingmann thanks for the feedback. I've added integration test. Original builds were failing because of wrong content of one of Python core libraries. My guess here would be some packaging issue.

The build is now passing after pushing new code that includes test.

jgehrcke commented:
I might be wrong, but from what I remember, previously the journal content for the `dcos-exhibitor` unit did not contain the contents that we found in `zookeeper.out`. Just to clarify: do we now expect the contents that we previously found in `zookeeper.out` to be accessible through
```
journaltcl -u dcos-exhibitor
```
? That is, does ZooKeeper (now) emit log messages to stdout, too?

gpaul commented:
The output is definitely different.

```
# zookeeper.log
2017-02-02 12:56:33,089 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:57574
2017-02-02 12:56:33,090 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing ruok command from /127.0.0.1:57574
2017-02-02 12:56:33,090 [myid:] - INFO  [Thread-56774:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:57574 (no session established for client)
2017-02-02 12:56:33,091 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:57576
2017-02-02 12:56:33,091 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing srvr command from /127.0.0.1:57576
2017-02-02 12:56:33,092 [myid:] - INFO  [Thread-56775:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:57576 (no session established for client)
2017-02-02 12:56:35,093 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:57604
2017-02-02 12:56:35,093 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing ruok command from /127.0.0.1:57604
2017-02-02 12:56:35,094 [myid:] - INFO  [Thread-56776:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:57604 (no session established for client)
2017-02-02 12:56:35,094 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:57606
2017-02-02 12:56:35,095 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing srvr command from /127.0.0.1:57606
2017-02-02 12:56:35,095 [myid:] - INFO  [Thread-56777:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:57606 (no session established for client)
```

```
# journalctl -f -u dcos-exhibitor
Feb 02 12:56:35 dcos-docker-master1 java[779]: Accepted socket connection from /127.0.0.1:57604
Feb 02 12:56:35 dcos-docker-master1 java[779]: Processing ruok command from /127.0.0.1:57604
Feb 02 12:56:35 dcos-docker-master1 java[779]: Closed socket connection for client /127.0.0.1:57604 (no session established for client)
Feb 02 12:56:35 dcos-docker-master1 java[779]: Accepted socket connection from /127.0.0.1:57606
Feb 02 12:56:35 dcos-docker-master1 java[779]: Processing srvr command from /127.0.0.1:57606
Feb 02 12:56:35 dcos-docker-master1 java[779]: Closed socket connection for client /127.0.0.1:57606 (no session established for client)

```

As you can see the console output (which gets redirected to zookeeper.out) seems like the same output we see in the journal with additional `[myid:] - INFO [Thread ...] - ` prefixed.

This seems to match the logging config in `/var/lib/dcos/exhibitor/conf/log4j.properties`:

```
log4j.rootLogger=INFO, journal, console

log4j.appender.journal=de.bwaldvogel.log4j.SystemdJournalAppender
log4j.appender.journal.logStacktrace=true
log4j.appender.journal.logThreadName=true
log4j.appender.journal.logLoggerName=true

log4j.appender.console=org.apache.log4j.ConsoleAppender
log4j.appender.console.Threshold=INFO
log4j.appender.console.layout=org.apache.log4j.PatternLayout
log4j.appender.console.layout.ConversionPattern=%d{ISO8601} [myid:%X{myid}] - %-5p [%t:%C{1}@%L] - %m%n
```

mhrabovcin commented:
Great catch @jgehrcke, @gpaul! Log format question led me to explore `log4j` options we pass to zookeeper and I've found a way to "cleanly" disable logging output to a file without need to link `zookeeper.out` file to `/dev/null`.

This patch now disables writes to `CONSOLE` log4j appender, which was redirected by `zkServer.sh` script to `zookeeper.out` file. I've also replaced `SystemdJournalAppender` with `SystemdJournalAppenderWithLayout` and changed default format of message to match format of messages written previously to `zookeeper.out` file.

Example of output:

```
Feb 02 14:40:32 dcos-docker-master1 java[379]: [Thread-890:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:51608 (no session established for client)
Feb 02 14:40:32 dcos-docker-master1 java[379]: [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:51609
Feb 02 14:40:32 dcos-docker-master1 java[379]: [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing srvr command from /127.0.0.1:51609
Feb 02 14:40:32 dcos-docker-master1 java[379]: [Thread-891:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:51609 (no session established for client)
Feb 02 14:40:33 dcos-docker-master1 java[379]: [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /127.0.0.1:51610
Feb 02 14:40:33 dcos-docker-master1 java[379]: [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@827] - Processing ruok command from /127.0.0.1:51610
Feb 02 14:40:33 dcos-docker-master1 java[379]: [Thread-892:NIOServerCnxn@1008] - Closed socket connection for client /127.0.0.1:51610 (no session established for client)
```

gpaul commented:
Great patch! Lovely solution.

jgehrcke commented:
Looks great! :rocket: 

> and changed default format of message to match format of messages written previously to zookeeper.out file.

Again, just to confirm.... That means we are now sure that we retain the same amount of information in the journal that we previously had in the `zookeeper.out` file?

mhrabovcin commented:
@jgehrcke correct, same information is begin written to journald and to stdout in very similar format.

lingmann commented:
Nice! The one bit of info we are not capturing, which was previously captured in zookeeper.out, appears to be the log4j log level (`%-5p`). Any value in capturing that as well in our journald logs?

mhrabovcin commented:
@lingmann the `SystemdJournalAppenderWithLayout` captures that information and sends it to `journald` but as part of journald binary message format and not as part of log text message. It would be easy to add though! Let me know if you'd like to have it as part of text message as well.

jgehrcke commented:
> appears to be the log4j log level (%-5p)

@mhrabovcin IMO the log level should be part of the message, yes. That takes some burden from the reader, and IIUC one can do something in the lines of ` journalctl -u dcos-exhibitor | grep WARNING` to filter for the level (for those who are too lazy to use the journald message data fields for that).

mhrabovcin commented:
@jgehrcke your use case makes sense. I've added severity level to log text message.

lingmann commented:
@mhrabovcin cool didn't realize that the log level was captured in a structured fashion. Nice work here! 🚢

jgehrcke commented:
This patch causes a regression described and debugged in https://jira.mesosphere.com/browse/DCOS_OSS-1301.
